sphinx_gallery/recommender.py

```diff
@@ -0,0 +1,273 @@
+
+
+class ExampleRecommender:
+    """Compute content-based KNN-TF-IFD recommendation system.
+
+    Parameters
+    ----------
+    n_examples : int, default=5
+        Number of most relevant examples to display.
+    min_df : float in range [0.0, 1.0] or int, default=1
+        When building the vocabulary ignore terms that have a document frequency
+        strictly lower than the given threshold. If float, the parameter
+        represents a proportion of documents, integer represents absolute
+        counts. This value is also called cut-off in the literature.
+    max_df : float in range [0.0, 1.0] or int, default=1.0
+        When building the vocabulary ignore terms that have a document frequency
+        strictly higher than the given threshold. If float, the parameter
+        represents a proportion of documents, integer represents absolute
+        counts.
+
+    Attributes
+    ----------
+    file_names_ : list of str
+        The list of file names used for computing the similarity matrix.
+        The recommended examples are chosen among this list.
+
+    similarity_matrix_ : dense matrix
+        Fitted matrix of pairwise cosine similarities.
+    """
+
+    def __init__(self, *, n_examples=5, min_df=3, max_df=0.9):
+        self.n_examples = n_examples
+        self.min_df = min_df
+        self.max_df = max_df
+
+    def token_freqs(self, doc):
+        """Extract a dict mapping raw tokens from doc to their occurrences."""
+        token_generator = (tok.lower() for tok in re.findall(r"\w+", doc))
+        return self.dict_freqs(token_generator)
+
+    @staticmethod
+    def dict_freqs(doc):
+        """Extract a dict mapping list of tokens to their occurrences."""
+        freq = defaultdict(int)
+        for tok in doc:
+            freq[tok] += 1
+        return freq
+
+    def dict_vectorizer(self, data):
+        """Convert a dictionary of feature arrays into a matrix.
+
+        Parameters
+        ----------
+        data : list of dict
+            An iterable of dictionaries of feature arrays, where each key
+            corresponds to a feature name, and each value is an array of
+            feature values.
+
+        Returns
+        -------
+        X : ndarray of shape (n_samples, n_features)
+            A matrix of occurrences where n_samples is the number of samples in
+            the dataset and n_features is the total number of features across
+            all samples.
+        """
+        try:
+            import numpy as np
+
+        except ImportError:
+            raise ConfigError("gallery_conf['recommender'] requires numpy")
+
+        feature_names = []
+        all_values = defaultdict(list)
+        for row in data:
+            for feature_name, feature_value in row.items():
+                feature_names.append(feature_name)
+                all_values[feature_name].append(feature_value)
+
+        feature_dict = {feature: i for i, feature in enumerate(sorted(all_values))}
```

Comment:
nitpick, `i` -> `col_idx` (or something to that effect)?

Reply:
Done in 494d177ffc695982e541a8bf45c7e2eb207fba77.